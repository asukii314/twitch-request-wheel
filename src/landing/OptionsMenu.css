#options-menu {}

#options-menu .options-menu-items .btn {
    display: block;
    text-align: left !important;
    font-size: calc(1.275rem + 0.3vw) !important;
    font-weight: 500 !important;
    opacity:0.8;
    transition: opacity 0.125s linear;
}
#options-menu .options-menu-items .btn:hover {
    opacity:1;
}
