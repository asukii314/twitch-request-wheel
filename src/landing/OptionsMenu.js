import {Component} from 'react';
import {Collapse, Offcanvas} from 'react-bootstrap';
import OptionsGameList from './OptionsGameList';
import PropTypes from 'prop-types';
import {version} from '../../package.json';

import './OptionsMenu.css';

export default class OptionsMenu extends Component {
    static get propTypes() {
        return {
            gamesList: PropTypes.object,
            onHide: PropTypes.func,
            onLogout: PropTypes.func,
            showOptionsMenu: PropTypes.bool
        };
    }
    static get defaultProps() {
        return {
            gamesList: {
                allowedGames: null,
                validGames: null
            },
            onHide: () => void 0,
            onLogout: () => void 0,
            showOptionsMenu: false
        };
    }

    constructor(props) {
        super(props);
        this.state = {
            showGameList: false
        }
        this.toggleGameList = this.toggleGameList.bind(this);
    }

    toggleGameList = () => {
        this.setState((state) => {
            return {
                showGameList: !state.showGameList
            }
        })
    }

    // renderGameOptions() {
    //     let {allowedGames, validGames} = this.props.gamesList;
    //     let gamePackList = [].concat(...Object.entries(validGames).map((packData, idx) => {
    //         return Object.keys(packData[1]).map(gameData => {
    //             let gameId = `${packData[0]} ${gameData}`.replace(/\W/ig, '_');
    //             return {
    //                 id: gameId,
    //                 game: gameData,
    //                 pack: packData[0]
    //             }
    //         })
    //     }))
    //
    //
    //
    //     // let gamesList = gamePackList.map(g => g.game);
    //     console.log('gamePackList:', gamePackList, allowedGames);
    //
    //     return (
    //         <Modal
    //             show={this.state.showOptionsModal}
    //             onHide={()=>this.toggleOptionsModal(false)}
    //             size="lg"
    //             aria-labelledby="contained-modal-title-vcenter"
    //             centered>
    //             <Modal.Header closeButton>
    //                 <Modal.Title id="contained-modal-title-vcenter">
    //                     Options
    //                 </Modal.Title>
    //             </Modal.Header>
    //             <Modal.Body>
    //                 <div className="options-list">
    //                     <ul>
    //                         {gamePackList.map(({id, game, pack}, idx) => {
    //                             // let gameId = `${g.pack} ${g.game}`.replace(/\W/ig, '_');
    //                             return (
    //                                 <li key={id}>
    //                                     <input type="checkbox" id={id} name={id} value={id} /> <label htmlFor={id}>{pack}: {game}</label>
    //                                 </li>
    //                             )}
    //                         )}
    //                     </ul>
    //                 </div>
    //             </Modal.Body>
    //             <Modal.Footer>
    //                 <Button data-bs-dismiss="modal">Close</Button>
    //             </Modal.Footer>
    //         </Modal>
    //     );
    // }

    render() {

        return (
            <Offcanvas
                id="options-menu"
                onHide={this.props.onHide}
                placement="end"
                show={this.props.showOptionsMenu}>
                <Offcanvas.Header closeButton closeVariant="white">
                    <Offcanvas.Title as="h2" className="fw-bold">
                        Options
                    </Offcanvas.Title>
                </Offcanvas.Header>
                <Offcanvas.Body>
                    <ul className="options-menu-items list-unstyled pb-3 px-4">
                        <li className="mb-1 fs-4 d-grid text-start">
                            <button className="btn logout" onClick={this.props.onLogout}>
                                Logout
                            </button>
                        </li>
                        <li className="mb-1 fs-4 d-grid text-start d-none">
                            <button className="btn game-list" onClick={this.toggleGameList}>
                                Game List
                            </button>
                        </li>
                        <Collapse in={this.state.showGameList}>
                            <div>
                                <OptionsGameList
                                    allowedGames={this.props.gamesList?.allowedGames}
                                    validGames={this.props.gamesList?.validGames} />
                            </div>
                        </Collapse>
                    </ul>


                    <div className="position-absolute bottom-0 start-50 translate-middle-x pb-3">
                        <small>{`version ${version}`}</small>
                    </div>
                </Offcanvas.Body>
            </Offcanvas>
        );
    }
}
