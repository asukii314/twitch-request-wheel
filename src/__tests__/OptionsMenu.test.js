import OptionsMenu from '../landing/OptionsMenu';
import {createRenderer} from 'react-test-renderer/shallow';
import React from 'react';

describe('OptionsMenu', () => {
    describe('initial state', () => {
        test('should return the base state object and default props', () => {
            let component = new OptionsMenu();
            let {state} = component;
            expect(state.showGameList).toBe(false);

            let props = OptionsMenu.defaultProps;
            expect(props.onHide()).toBeUndefined();
            expect(props.onLogout()).toBeUndefined();
            expect(props.showOptionsMenu).toBe(false);
        });
    });
    describe('createDebugMenuItems', () => {
        test('should return array of items', () => {
            const component = new OptionsMenu();
            let items = [{
                label: 'item',
                onClick: jest.fn()
            }];
            let menuItems = component.createDebugMenuItems(items);
            expect(menuItems).toMatchSnapshot();
        });
        test('should return an empty array', () => {
            const component = new OptionsMenu();
            expect(component.createDebugMenuItems()).toEqual([]);
        });
    });
    describe('createMenuItems', () => {
        test('should return array of items', () => {
            const component = new OptionsMenu();
            let items = [{
                label: 'item',
                onClick: jest.fn()
            }];
            let menuItems = component.createMenuItems(items);
            expect(menuItems).toMatchSnapshot();
        });
        test('should return an empty array', () => {
            const component = new OptionsMenu();
            expect(component.createMenuItems()).toEqual([]);
        });
    });
    describe('toggleGameList', () => {
        test('should change state.showGameList', () => {
            const component = new OptionsMenu();
            jest.spyOn(component, 'setState').mockImplementation(()=>{});

            component.toggleGameList();

            expect(component.setState).toHaveBeenCalledTimes(1);
            expect(component.setState.mock.calls[0][0]({})).toEqual({showGameList: true});
        });
    });
    describe('render', () => {
        test('should render without error', () => {
            const shallowRenderer = createRenderer();
            shallowRenderer.render(<OptionsMenu />);

            let component = shallowRenderer.getRenderOutput();
            expect(component).not.toBe(null);
            expect(component).toMatchSnapshot();
            shallowRenderer.unmount();
        });
        test('should render with menu items', () => {
            const props = {
                showOptionsMenu: true
            };
            const shallowRenderer = createRenderer();
            shallowRenderer.render(<OptionsMenu {...props} />);
            let instance = shallowRenderer.getMountedInstance();
            instance.setState({
                showGameList: true
            });
            let component = shallowRenderer.getRenderOutput();
            expect(component).not.toBe(null);
            expect(component).toMatchSnapshot();
            shallowRenderer.unmount();
        });
    });
});
