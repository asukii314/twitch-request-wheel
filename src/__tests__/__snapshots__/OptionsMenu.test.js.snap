// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionsMenu createDebugMenuItems should return array of items 1`] = `
Array [
  <DropdownItem
    eventKey="item"
    href={null}
    onClick={[MockFunction]}
  >
    item
  </DropdownItem>,
  <DropdownDivider
    role="separator"
  />,
  <DropdownItem
    eventKey="item"
    href={null}
    onClick={null}
  >
    item
  </DropdownItem>,
]
`;

exports[`OptionsMenu createDebugMenuItems should return array of remaining items when some items are missing keys or values 1`] = `
Array [
  <DropdownItem
    eventKey="item"
    href={null}
    onClick={[MockFunction]}
  >
    item
  </DropdownItem>,
]
`;

exports[`OptionsMenu createMenuItems should return array of items and filter invalid items 1`] = `
Array [
  <li
    className="mb-1 fs-4 d-grid text-start item1"
  >
    <Button
      active={false}
      className="btn"
      disabled={false}
      onClick={[Function]}
      variant="link"
    >
      item1
    </Button>
  </li>,
  <li
    className="mb-1 fs-4 d-grid text-start"
  >
    <Button
      active={false}
      className="btn"
      disabled={false}
      onClick={[MockFunction]}
      variant="link"
    >
       
    </Button>
  </li>,
  <li
    className="mb-1 fs-4 d-grid text-start my-item"
  >
    <Button
      active={false}
      className="btn"
      disabled={false}
      onClick={[MockFunction]}
      variant="link"
    >
      item3
    </Button>
  </li>,
]
`;

exports[`OptionsMenu render should render with menu items 1`] = `
<Offcanvas
  autoFocus={true}
  backdrop={true}
  enforceFocus={true}
  id="options-menu"
  keyboard={true}
  onHide={[Function]}
  placement="end"
  renderStaticNode={false}
  restoreFocus={true}
  scroll={false}
  show={true}
>
  <OffcanvasHeader
    closeButton={true}
    closeLabel="Close"
    closeVariant="white"
  >
    <OffcanvasTitle
      as="h2"
      className="fw-bold"
    >
      Options
    </OffcanvasTitle>
  </OffcanvasHeader>
  <OffcanvasBody>
    <ul
      className="options-menu-items list-unstyled pb-3 px-4"
    >
      <li
        className="mb-1 fs-4 d-grid text-start"
      >
        <Button
          active={false}
          className="btn logout"
          disabled={false}
          onClick={[Function]}
          variant="link"
        >
          Logout
        </Button>
      </li>
      <hr />
      <li
        className="mb-1 fs-4 d-grid text-start"
      >
        <Button
          active={false}
          className="btn undo-start-game"
          disabled={false}
          onClick={[Function]}
          variant="link"
        >
          Undo Start
          <div
            className="fs-6 fst-italic opacity-75 lh-1"
          >
            Start Time: 
          </div>
        </Button>
      </li>
      <li
        className="mb-1 fs-4 d-grid text-start"
      >
        <Button
          active={false}
          className="btn settings-menu"
          disabled={false}
          onClick={[Function]}
          variant="link"
        >
          Settings
        </Button>
      </li>
      <ForwardRef
        appear={false}
        getDimensionValue={[Function]}
        in={false}
        mountOnEnter={false}
        timeout={300}
        unmountOnExit={false}
      >
        <div
          className="accordion-dark accordion accordion-flush"
          id="settings-menu"
        >
          <div
            className="accordion-body"
          >
            <Button
              active={false}
              className="btn settings-menu"
              disabled={false}
              onClick={[Function]}
              title="Allows subscribers to make additional game requests when enabled."
              variant="link"
            >
              <input
                readOnly={true}
                role="switch"
                type="checkbox"
              />
               
              <span>
                Enable Sub Requests
              </span>
            </Button>
            <Button
              active={false}
              className="btn settings-menu subsetting"
              disabled={true}
              onClick={[Function]}
              title="Limit subscribers to one additional game requests when enabled."
              variant="link"
            >
              <input
                readOnly={true}
                role="switch"
                type="checkbox"
              />
               
              <span>
                Limit 1 Sub Request
              </span>
            </Button>
            <Button
              active={false}
              className="btn settings-menu"
              disabled={false}
              onClick={[Function]}
              title="Clears the list of player signups after a game redemption."
              variant="link"
            >
              <input
                readOnly={true}
                role="switch"
                type="checkbox"
              />
               
              <span>
                Clear Seats After Redeem
              </span>
            </Button>
            <Button
              active={false}
              className="btn settings-menu"
              disabled={false}
              title="Uses a custom character or emote to separate requests listed in the chat."
              variant="link"
            >
              <span>
                Use Custom Delimiter: 
              </span>
              <input
                className="form-control"
                name="custom-delimiter"
                onChange={[Function]}
                type="text"
              />
            </Button>
            <Button
              active={false}
              className="btn settings-menu"
              disabled={false}
              onClick={[Function]}
              title="Allows host to set a room code that can be whispered to players."
              variant="link"
            >
              <input
                readOnly={true}
                role="switch"
                type="checkbox"
              />
               
              <span>
                Enable Room Code
              </span>
            </Button>
            <Button
              active={false}
              className="btn settings-menu"
              disabled={false}
              onClick={[Function]}
              title="Posts a link in chat to the commands page when '!command' is used; otherwise, lists all available commands in chat when disabled."
              variant="link"
            >
              <input
                readOnly={true}
                role="switch"
                type="checkbox"
              />
               
              <span>
                Use Link for 
                <tt>
                  !command
                </tt>
              </span>
            </Button>
          </div>
        </div>
      </ForwardRef>
      <li
        className="mb-1 fs-4 d-grid text-start d-none"
      >
        <Button
          active={false}
          className="btn reload-game-list"
          disabled={false}
          variant="link"
        >
          Refresh Game List
        </Button>
      </li>
      <li
        className="mb-1 fs-4 d-grid text-start d-none"
      >
        <Button
          active={false}
          className="btn game-list"
          disabled={false}
          onClick={[Function]}
          variant="link"
        >
          Game List
        </Button>
      </li>
      <ForwardRef
        appear={false}
        getDimensionValue={[Function]}
        in={true}
        mountOnEnter={false}
        timeout={300}
        unmountOnExit={false}
      >
        <div>
          <OptionsGameList
            allowedGames={null}
            validGames={null}
          />
        </div>
      </ForwardRef>
    </ul>
    <div
      className="position-absolute bottom-0 start-0 end-0 pb-3 text-center"
      id="options-debug-menu-items"
    >
      <Dropdown
        align="start"
        autoClose={true}
        drop="up-centered"
        id="dropdown-debug-menu-items"
        navbar={false}
        variant="link"
      >
        <DropdownToggle
          id="dropdown-debug-menu-items-toggle"
          size="sm"
          variant="link"
        >
          version 0.0.0
        </DropdownToggle>
        <DropdownMenu
          flip={true}
          variant="dark"
        >
          <DropdownHeader
            role="heading"
          >
            Debug Options
          </DropdownHeader>
        </DropdownMenu>
      </Dropdown>
    </div>
  </OffcanvasBody>
</Offcanvas>
`;

exports[`OptionsMenu render should render without error 1`] = `
<Offcanvas
  autoFocus={true}
  backdrop={true}
  enforceFocus={true}
  id="options-menu"
  keyboard={true}
  onHide={[Function]}
  placement="end"
  renderStaticNode={false}
  restoreFocus={true}
  scroll={false}
  show={false}
>
  <OffcanvasHeader
    closeButton={true}
    closeLabel="Close"
    closeVariant="white"
  >
    <OffcanvasTitle
      as="h2"
      className="fw-bold"
    >
      Options
    </OffcanvasTitle>
  </OffcanvasHeader>
  <OffcanvasBody>
    <ul
      className="options-menu-items list-unstyled pb-3 px-4"
    >
      <li
        className="mb-1 fs-4 d-grid text-start"
      >
        <Button
          active={false}
          className="btn logout"
          disabled={false}
          onClick={[Function]}
          variant="link"
        >
          Logout
        </Button>
      </li>
      <hr />
      <li
        className="mb-1 fs-4 d-grid text-start"
      >
        <Button
          active={false}
          className="btn settings-menu"
          disabled={false}
          onClick={[Function]}
          variant="link"
        >
          Settings
        </Button>
      </li>
      <ForwardRef
        appear={false}
        getDimensionValue={[Function]}
        in={false}
        mountOnEnter={false}
        timeout={300}
        unmountOnExit={false}
      >
        <div
          className="accordion-dark accordion accordion-flush"
          id="settings-menu"
        >
          <div
            className="accordion-body"
          >
            <Button
              active={false}
              className="btn settings-menu"
              disabled={false}
              onClick={[Function]}
              title="Allows subscribers to make additional game requests when enabled."
              variant="link"
            >
              <input
                readOnly={true}
                role="switch"
                type="checkbox"
              />
               
              <span>
                Enable Sub Requests
              </span>
            </Button>
            <Button
              active={false}
              className="btn settings-menu subsetting"
              disabled={true}
              onClick={[Function]}
              title="Limit subscribers to one additional game requests when enabled."
              variant="link"
            >
              <input
                readOnly={true}
                role="switch"
                type="checkbox"
              />
               
              <span>
                Limit 1 Sub Request
              </span>
            </Button>
            <Button
              active={false}
              className="btn settings-menu"
              disabled={false}
              onClick={[Function]}
              title="Clears the list of player signups after a game redemption."
              variant="link"
            >
              <input
                readOnly={true}
                role="switch"
                type="checkbox"
              />
               
              <span>
                Clear Seats After Redeem
              </span>
            </Button>
            <Button
              active={false}
              className="btn settings-menu"
              disabled={false}
              title="Uses a custom character or emote to separate requests listed in the chat."
              variant="link"
            >
              <span>
                Use Custom Delimiter: 
              </span>
              <input
                className="form-control"
                name="custom-delimiter"
                onChange={[Function]}
                type="text"
              />
            </Button>
            <Button
              active={false}
              className="btn settings-menu"
              disabled={false}
              onClick={[Function]}
              title="Allows host to set a room code that can be whispered to players."
              variant="link"
            >
              <input
                readOnly={true}
                role="switch"
                type="checkbox"
              />
               
              <span>
                Enable Room Code
              </span>
            </Button>
            <Button
              active={false}
              className="btn settings-menu"
              disabled={false}
              onClick={[Function]}
              title="Posts a link in chat to the commands page when '!command' is used; otherwise, lists all available commands in chat when disabled."
              variant="link"
            >
              <input
                readOnly={true}
                role="switch"
                type="checkbox"
              />
               
              <span>
                Use Link for 
                <tt>
                  !command
                </tt>
              </span>
            </Button>
          </div>
        </div>
      </ForwardRef>
      <li
        className="mb-1 fs-4 d-grid text-start d-none"
      >
        <Button
          active={false}
          className="btn reload-game-list"
          disabled={false}
          variant="link"
        >
          Refresh Game List
        </Button>
      </li>
      <li
        className="mb-1 fs-4 d-grid text-start d-none"
      >
        <Button
          active={false}
          className="btn game-list"
          disabled={false}
          onClick={[Function]}
          variant="link"
        >
          Game List
        </Button>
      </li>
      <ForwardRef
        appear={false}
        getDimensionValue={[Function]}
        in={false}
        mountOnEnter={false}
        timeout={300}
        unmountOnExit={false}
      >
        <div>
          <OptionsGameList
            allowedGames={null}
            validGames={null}
          />
        </div>
      </ForwardRef>
    </ul>
    <div
      className="position-absolute bottom-0 start-0 end-0 pb-3 text-center"
      id="options-debug-menu-items"
    >
      <Dropdown
        align="start"
        autoClose={true}
        drop="up-centered"
        id="dropdown-debug-menu-items"
        navbar={false}
        variant="link"
      >
        <DropdownToggle
          id="dropdown-debug-menu-items-toggle"
          size="sm"
          variant="link"
        >
          version 0.0.0
        </DropdownToggle>
        <DropdownMenu
          flip={true}
          variant="dark"
        >
          <DropdownHeader
            role="heading"
          >
            Debug Options
          </DropdownHeader>
        </DropdownMenu>
      </Dropdown>
    </div>
  </OffcanvasBody>
</Offcanvas>
`;
